### ISSUE TITLE
SVM Unexpected Performance

### ISSUE THREAD

**Issue by ryuga93** [kyamagu/mexopencv#243]
Hello, I am currrently doing Kernel PCA with SVM for face recognition and I am using the PCA code from the matlab community:
[http://www.mathworks.com/matlabcentral/fileexchange/39715-kernel-pca-and-pre-image-reconstruction](http://www.mathworks.com/matlabcentral/fileexchange/39715-kernel-pca-and-pre-image-reconstruction)

The problems I am currently facing is Matlab will close itself when I am using SVM/train() with high poly degree(3 and above) and could not write to XML file when the PCA poly is 3 or above with this error:
Unexpected Standard exception from MEX file.
```
What() is:C:\builds\master_PackSlave-win64-vc12-shared\opencv\modules\ml\src\svm.cpp:2038: error: (-212) SVM model data is
invalid, check sv_count, var_* and class_count tags in function cv::ml::SVMImpl::write
```

**Comment by amroamroamro** [comment]
I can't reproduce the error. Here is a simple test:

```matlab
X = [randn(50,3)+1; randn(50,3)-1];
Y = int32([ones(50,1); -ones(50,1)]);
model = cv.SVM();
model.Type = 'C_SVC';
model.KernelType = 'Poly';
model.Degree = 5;
model.train(X,Y);
out = model.save('.yml')
```

I get normal results for both XML or YML files:

``` yml
>> out
out =
%YAML:1.0
opencv_ml_svm:
   format: 3
   svmType: C_SVC
   kernel:
      type: POLY
      degree: 5.
      gamma: 1.
      coef0: 0.
   C: 1.
   term_criteria: { epsilon:1.1920928955078125e-07, iterations:1000 }
   var_count: 3
   class_count: 2
   class_labels: !!opencv-matrix
      rows: 2
      cols: 1
      dt: i
      data: [ -1, 1 ]
   sv_total: 15
   support_vectors:
      - [ -1.20647478e+00, 1.70146966e+00, 9.55046952e-01 ]
      - [ 4.44634289e-01, 4.53069359e-01, 1.46405131e-01 ]
      - [ 1.40720397e-01, 2.41152358e+00, -1.16633856e+00 ]
      - [ 6.91993773e-01, -1.52452314e+00, -1.57974973e-01 ]
      - [ 1.86273503e+00, -6.28384948e-03, 1.20139062e+00 ]
      - [ 6.60645902e-01, 3.38792890e-01, -9.76718724e-01 ]
      - [ 2.11010957e+00, 8.80508184e-01, -3.51432562e-01 ]
      - [ -2.10270405e+00, -6.47465214e-02, 1.56205678e+00 ]
      - [ 9.44574252e-02, 1.08114338e+00, -7.76994050e-01 ]
      - [ -1.60638261e+00, -6.98868513e-01, 6.55608438e-03 ]
      - [ -1.21575856e+00, 4.73418981e-01, 6.56834960e-01 ]
      - [ -1.34826541e+00, -4.00835335e-01, -6.61312938e-02 ]
      - [ 6.28963485e-02, 5.38209192e-02, -1.55770564e+00 ]
      - [ 8.70702386e-01, 1.38987936e-02, 3.48327532e-02 ]
      - [ 1.48267698e+00, -2.07794285e+00, -1.17626953e+00 ]
   decision_functions:
      -
         sv_count: 15
         rho: 9.1196728306453423e-02
         alpha: [ 2.4710137419353614e-01, 2.9076447102233669e-04,
             1.4607739941539860e-01, 2.0450507644703053e-03,
             3.4643462342999379e-02, 3.3874934513956989e-01,
             3.3825869312226692e-03, -1.3870656556536627e-03,
             -4.0797695404302625e-01, -7.2362795567962766e-03,
             -3.7832483828792049e-02, -3.5595753270749878e-04,
             -3.0507342858023823e-01, -3.2367687739775089e-03,
             -9.1910452870282693e-03 ]
```

**Comment by ryuga93** [comment]
I think the model will "breakdown" when the numbers is large, but that's is my assumption. Maybe you could try to train the model using my data:
[https://drive.google.com/file/d/0B9f-KP4kYSk-cTJ3ZnRZdS1wdGM/view?usp=sharing](https://drive.google.com/file/d/0B9f-KP4kYSk-cTJ3ZnRZdS1wdGM/view?usp=sharing)

Please try to train a model with _degree 3 or above_ using _weight1.mat_ as _sample_ and _target.mat_ as _response_. The Matlab will close itself in this case.

Then try train a model with _degree 1_ using _weight2.mat_ as _sample_ and _target.mat_ as _response_, then use model.save() to produce the XML file. In this case, I got the save error mentioned and the XML file only contains this information:
`<?xml version="1.0"?>
-<opencv_storage>
-<opencv_ml_svm>
<format>3</format>
</opencv_ml_svm>
</opencv_storage>`

**Comment by amroamroamro** [comment]
Ok I checked your data, and I see the problem.. Try this code to understand:

``` matlab
load weight1.mat
load target.mat
X = weight;
Y = target;
model = cv.SVM();
model.Type = 'C_SVC';
model.KernelType = 'Poly';
model.Degree = 3;
if ~model.train(X,Y)
    error('training failed')
end
if model.isTrained()
    out = model.save('.yml')
end

```

The training actually failed (it returns false), so the model is cleared and empty on return, which is why it throws a C++ exception when you try to save the empty model. It's like doing this:

``` matlab
model = cv.SVM();
model.save('out.xml')
```

So this is not a bug.

---

Now the reason that the training part is failing probably has to do with your data. I see you have 140 samples each with 91 features, grouped into 28 classes with 5 samples per class. This is just not enough data for that many classes and high dimensions (recall that SVM trains one-vs-one `k*(k-1)/2` binary models for multiclass problems), especially when you increase the model complexity (polynomial degree)...

More importantly, SVM classifiers are quite sensitive to the scale of features. The first few features you have are very large compared to later ones (most of the data is in the order of 1e8):

```matlab
>> parallelcoords(weight, 'Quantile',0.25)
```

![parallelcoords](http://i.stack.imgur.com/0ewCi.png)

Consider normalizing the data range before training:

``` matlab
X = zscore(weight);
success = model.train(X,Y);
```

This should work even for `degree=3` or higher.

**Comment by amroamroamro** [comment]
To be more exact, using the debugger I traced it to this section of the code:
https://github.com/Itseez/opencv/blob/3.1.0/modules/ml/src/svm.cpp#L686

You can see it checks if intermediate values became too big > 1e300, and returns false if so. Note that this sanity check only happens in debug mode (it's surrounded by `#ifdef _DEBUG`), so in release mode those checks are skipped and further calculations will likely result in `Inf` or `NaN` values causing the crash somewhere later...

**Comment by ryuga93** [comment]
I see, thank you very much for answering this clearly. Looks like I need to read more about SVM...I would like to know does your code above crashes? Actually I am using GUI and I have the training status check before saving the model
```
status=model.train(weight,target);
if (status==1)
   msgbox('Success!','Train SVM','none','modal'); 
   model.save([path file]);
else
    errordlg('Error occurred!','Train SVM,','modal');
end
```

**Comment by amroamroamro** [comment]
Unfortunately yes, it will crash running with `weight1.mat` in **release mode** (debug mode actually handles the data more gracefully returning false). Like I explained in the above comment, that section of the code which checks the values is only performed in debug mode.

Ultimately this is an issue on the OpenCV not mexopencv, consider filing a bug report to OpenCV, maybe the checks should be done for both debug and release...

Either way, scaling the data should now solve the problem for you.

**Comment by ryuga93** [comment]
Alright, thank you.
